lib/internal/modules/esm/hooks.js

```diff
@@ -114,6 +114,25 @@ class Hooks {
     this.addCustomLoaders(userLoaders);
   }
 
+  /**
+   * Import and register custom/user-defined module loader hook(s).
+   * @param {string} urlOrSpecifier
+   */
+  async registerLoader(urlOrSpecifier) {
```

Comment:
this should throw if it's not run in the main thread

Reply:
You mean if user application code runs it in a worker? Why wouldn’t this be available everywhere?

Reply:
If this is called in a different thread, it would be the user expectation that the new loader would apply only to that thread. This will require spinning up a new loader thread etc, falling back into the same problem of `--experimental-loader` in `execArgv`.

Reply:
I believe we explicitly don't want multiple loader threads

Reply:
I don’t see the justification for being able to customize loaders per thread. That’s not possible via `require.extensions` as far as I’m aware. Our intent was to have one loaders thread that handles all user application threads, both main and workers. If a different Node configuration is needed for a subset of an application, that subset can be run in a child process.

Reply:
`require()` is configurable on each thread:

```js
'use strict';
const { Worker } = require('worker_threads');

require.extensions['.sjs'] = require.extensions['.js'];
console.log(require.extensions);
console.log(require.cache);

new Worker(`
  console.log(require.extensions);
  console.log(require.cache);
  `, { eval: true });
```

Reply:
Even still, I don’t see why worker threads need independent loader threads. We could potentially have the current behavior as a default that could be overridden when a worker thread is created, but I still strain to see why such a feature would be worth the additional complexity it would entail on our part and on the part of loader authors.

Reply:
So, I still think that being able to spin up a separate loader thread for any Worker that wants one is not a big deal - yes it consumes more resources, but people would be opting into that behavior.

That said, if having a single loaders thread is an explicit goal, would it be possible to make the loader thread aware of who is invoking it? That way, a single thread could intelligently multiplex between the main thread and worker threads if the user wanted that. The userland loaders themselves wouldn't need to be aware of that.

Reply:
I have no interest in such a feature, and I have a feeling that none of the people who just spent eight months trying to get a single loaders thread working would be eager to dive back into threads management to try to get something so complex to work. If you want to take a stab at it, be my guest, but for me the user story needs to be very compelling to add even more complexity onto something that’s already frightfully complex.

Reply:
I also don't need such a feature. `registerLoader()` is enough for me. I do think it's important to call out differences like this though. Presumably, people want ESM to be adopted, but it still makes so many things extremely complicated that are simple to do with CJS (module reloading, mocking, monkey patching, etc.). I think fewer of those discrepancies would push many more people toward ESM.

Reply:
FYI @cjihrig: https://github.com/nodejs/loaders/issues/140#issuecomment-1537225149.

I was mistaken in my understanding of how things currently work since we landed the off-thread PR. According to @targos the current behavior is that a new loaders thread is spawned for each new `Worker` thread, and the new loaders thread inherits whatever options were defined when the original process was launched (i.e. whatever was defined initially via `--loader` flags, ignoring whatever might be passed to `new Worker`). And like you saw, trying to define different loaders for a particular new loaders thread is currently impossible.

My and @JakobJingleheimer’s intention with the off-thread PR was that there should be only _one_ loaders thread regardless of how many application threads there are. I think this is the architecture that would be most likely to be desired by users, as it provides the best performance for two reasons:

- Typically a single loaders thread should be all that is needed, because it’s less common to want differing configurations of loaders for different worker threads, and so therefore for the common case we don’t want to spawn more threads than are necessary.

- Having multiple loaders threads probably causes increased and unnecessary work to be done, as each loaders thread would get its own scope. A transpiler loader, for example, might re-transpile the same file multiple times, one for each loaders thread, or have the overhead of needing to communicate to other threads to check if a transpiled file exists in a shared cache.

That’s why I feel like having only one loaders thread regardless of the number of application threads feels like the correct default behavior. That said, if someone wants to put in the effort, we could make the “new loaders thread to accompany a `Worker` thread” behavior something that can be requested as part of the `new Worker` options. I would be fine with that, as it would mean the user is opting into the less-performant architecture. But first I think we need to get to the “single loaders thread all the time” version, as I think that we should default to the most performant architecture.